src/MongoDB.Driver.Core/Core/Servers/ServerDescription.cs

```diff
@@ -507,57 +507,27 @@ public override string ToString()
-                    canonicalEndPoint: canonicalEndPoint.WithDefault(_canonicalEndPoint),
-                    electionId: electionId.WithDefault(_electionId),
-                    heartbeatException: heartbeatException.WithDefault(_heartbeatException),
-                    heartbeatInterval: heartbeatInterval.WithDefault(_heartbeatInterval),
-                    lastUpdateTimestamp: lastUpdateTimestamp.WithDefault(_lastUpdateTimestamp),
-                    lastWriteTimestamp: lastWriteTimestamp.WithDefault(_lastWriteTimestamp),
-                    logicalSessionTimeout: logicalSessionTimeout.WithDefault(_logicalSessionTimeout),
-                    maxBatchCount: maxBatchCount.WithDefault(_maxBatchCount),
-                    maxDocumentSize: maxDocumentSize.WithDefault(_maxDocumentSize),
-                    maxMessageSize: maxMessageSize.WithDefault(_maxMessageSize),
-                    maxWireDocumentSize: maxWireDocumentSize.WithDefault(_maxWireDocumentSize),
-                    replicaSetConfig: replicaSetConfig.WithDefault(_replicaSetConfig),
-                    state: state.WithDefault(_state),
-                    tags: tags.WithDefault(_tags),
-                    type: type.WithDefault(_type),
-                    version: version.WithDefault(_version),
-                    wireVersionRange: wireVersionRange.WithDefault(_wireVersionRange));
-            }
-            else
-            {
-                return this;
-            }
+            return new ServerDescription(
+                _serverId,
+                _endPoint,
+                averageRoundTripTime: averageRoundTripTime.WithDefault(_averageRoundTripTime),
+                canonicalEndPoint: canonicalEndPoint.WithDefault(_canonicalEndPoint),
+                electionId: electionId.WithDefault(_electionId),
+                heartbeatException: heartbeatException.WithDefault(_heartbeatException),
+                heartbeatInterval: heartbeatInterval.WithDefault(_heartbeatInterval),
+                lastUpdateTimestamp: lastUpdateTimestamp.WithDefault(DateTime.UtcNow),
+                lastWriteTimestamp: lastWriteTimestamp.WithDefault(_lastWriteTimestamp),
+                logicalSessionTimeout: logicalSessionTimeout.WithDefault(_logicalSessionTimeout),
+                maxBatchCount: maxBatchCount.WithDefault(_maxBatchCount),
+                maxDocumentSize: maxDocumentSize.WithDefault(_maxDocumentSize),
+                maxMessageSize: maxMessageSize.WithDefault(_maxMessageSize),
+                maxWireDocumentSize: maxWireDocumentSize.WithDefault(_maxWireDocumentSize),
+                replicaSetConfig: replicaSetConfig.WithDefault(_replicaSetConfig),
+                state: state.WithDefault(_state),
+                tags: tags.WithDefault(_tags),
+                type: type.WithDefault(_type),
+                version: version.WithDefault(_version),
+                wireVersionRange: wireVersionRange.WithDefault(_wireVersionRange));
```

Comment:
Other potential changes to ServerDescription:

1. In master at line 462 on the left we add "LastUpdateTimeStamp" to the ToString representation.

2. In master at line 551 on the left we use "DateTime.UtcNow" instead of "_lastUpdateTimeStamp" in the WithHeartbeatException.

These changes are probably worth including since this ticket is about improving server selection reliability and it would be helpful to have accurate timestamps when diagnosing future remaining problem with server selection.

@jyemin should weigh in on whether he wants these two suggested changes.

Reply:
When you say "in master", do you mean before or after the other PR (the PR against master) for this ticket was merged?

Reply:
I mean what's in the tip of master right now, which currently is:

https://github.com/mongodb/mongo-csharp-driver/commit/93075deffcbb819658fe51c98fdcd44906efd53b

So that includes the changes made in master for CSHARP-2565.

Reply:
@vincentkam over to you.  I'm not clear on why the two patches differ in this regard.

Reply:
I dug a bit deeper and found that the two lines I mentioned above were actually changed as part of working on different tickets:

1. https://github.com/mongodb/mongo-csharp-driver/commit/dd6d09b195cdd521cf4a1ab754d417c77b3273ac

2. https://github.com/mongodb/mongo-csharp-driver/commit/3a9d718e529b9e6dc658efed930d31f02a337e0b

So maybe we just backport those also?

Reply:
Nice catch Robert :) I'm fine with backporting those two if @jyemin is

Reply:
Sure, we can backport those.  But that can be done separately, as they are different tickets.